{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "063947c6-bb14-491c-aef0-7febf7ab5b60",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ozlemtuna/miniconda3/envs/jupyter_env/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/home/ozlemtuna/miniconda3/envs/jupyter_env/lib/python3.10/site-packages/docrep/decorators.py:43: SyntaxWarning: 'param_categorical_covariate_keys' is not a valid key!\n",
      "  doc = func(self, args[0].__doc__, *args[1:], **kwargs)\n",
      "/home/ozlemtuna/miniconda3/envs/jupyter_env/lib/python3.10/site-packages/docrep/decorators.py:43: SyntaxWarning: 'param_continuous_covariate_keys' is not a valid key!\n",
      "  doc = func(self, args[0].__doc__, *args[1:], **kwargs)\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import cell2location\n",
    "import scvi\n",
    "import torch\n",
    "import matplotlib as mpl\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib.backends.backend_pdf as mpdf\n",
    "import warnings\n",
    "sys.path.append(\"/home/biolab/Projects/LAMs_wd/scripts\")\n",
    "warnings.filterwarnings('ignore')\n",
    "data_type = 'float32'\n",
    "from visium_qc_and_visualisation import read_and_qc\n",
    "from config import lung_config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4c07d2ba-9215-4955-bc1a-ecc9df1574da",
   "metadata": {},
   "outputs": [],
   "source": [
    "env = \"Tumour\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bba1b527-a1aa-4129-88c8-1af771e2bd32",
   "metadata": {},
   "source": [
    "Load the signature matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fdd70584-eaea-4617-98d9-def067f170f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Stromal</th>\n",
       "      <th>Endothelial_cell</th>\n",
       "      <th>Alveolar_cell_type_1</th>\n",
       "      <th>Alveolar_cell_type_2</th>\n",
       "      <th>transitional_club/AT2</th>\n",
       "      <th>other</th>\n",
       "      <th>Macrophage_alveolar</th>\n",
       "      <th>NK_cell</th>\n",
       "      <th>Mast_cell</th>\n",
       "      <th>Club</th>\n",
       "      <th>...</th>\n",
       "      <th>T_cell_CD8</th>\n",
       "      <th>Tumor_cells</th>\n",
       "      <th>T_cell_CD4</th>\n",
       "      <th>Plasma_cell</th>\n",
       "      <th>cDC1</th>\n",
       "      <th>cDC2</th>\n",
       "      <th>B_cell</th>\n",
       "      <th>Neutrophils</th>\n",
       "      <th>DC_mature</th>\n",
       "      <th>T_cell_regulatory</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDH11</th>\n",
       "      <td>429.229015</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOCK9</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>473.570096</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>59.077085</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLIC3</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>997.781399</td>\n",
       "      <td>212.359701</td>\n",
       "      <td>237.847267</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>383.494818</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCNG2</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>102.533423</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>52.746868</td>\n",
       "      <td>56.873905</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>314.848934</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCDH17</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>560.654133</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDKN3</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>262.356384</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>93.431375</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAMK4</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>175.347016</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>213.361118</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>106.219895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WIF1</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>209.595993</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM156</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>76.778009</td>\n",
       "      <td>119.800078</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>167.259363</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLCO5A1</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>272.951785</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1149 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Stromal  Endothelial_cell  Alveolar_cell_type_1  \\\n",
       "CDH11    429.229015          1.000000              1.000000   \n",
       "DOCK9      1.000000        473.570096              1.000000   \n",
       "CLIC3      1.000000          1.000000            997.781399   \n",
       "CCNG2      1.000000          1.000000              1.000000   \n",
       "PCDH17     1.000000        560.654133              1.000000   \n",
       "...             ...               ...                   ...   \n",
       "CDKN3      1.000000          1.000000              1.000000   \n",
       "CAMK4      1.000000          1.000000              1.000000   \n",
       "WIF1       1.000000          1.000000              1.000000   \n",
       "TMEM156    1.000000          1.000000              1.000000   \n",
       "SLCO5A1    1.000000          1.000000              1.000000   \n",
       "\n",
       "         Alveolar_cell_type_2  transitional_club/AT2       other  \\\n",
       "CDH11                1.000000               1.000000    1.000000   \n",
       "DOCK9               59.077085               1.000000    1.000000   \n",
       "CLIC3              212.359701             237.847267    1.000000   \n",
       "CCNG2                1.000000               1.000000    1.000000   \n",
       "PCDH17               1.000000               1.000000    1.000000   \n",
       "...                       ...                    ...         ...   \n",
       "CDKN3                1.000000               1.000000  262.356384   \n",
       "CAMK4                1.000000               1.000000    1.000000   \n",
       "WIF1               209.595993               1.000000    1.000000   \n",
       "TMEM156              1.000000               1.000000    1.000000   \n",
       "SLCO5A1              1.000000               1.000000    1.000000   \n",
       "\n",
       "         Macrophage_alveolar     NK_cell   Mast_cell  Club  ...  T_cell_CD8  \\\n",
       "CDH11                    1.0    1.000000    1.000000   1.0  ...    1.000000   \n",
       "DOCK9                    1.0    1.000000    1.000000   1.0  ...    1.000000   \n",
       "CLIC3                    1.0  383.494818    1.000000   1.0  ...    1.000000   \n",
       "CCNG2                    1.0    1.000000  102.533423   1.0  ...    1.000000   \n",
       "PCDH17                   1.0    1.000000    1.000000   1.0  ...    1.000000   \n",
       "...                      ...         ...         ...   ...  ...         ...   \n",
       "CDKN3                    1.0    1.000000    1.000000   1.0  ...    1.000000   \n",
       "CAMK4                    1.0    1.000000    1.000000   1.0  ...  175.347016   \n",
       "WIF1                     1.0    1.000000    1.000000   1.0  ...    1.000000   \n",
       "TMEM156                  1.0    1.000000    1.000000   1.0  ...    1.000000   \n",
       "SLCO5A1                  1.0    1.000000    1.000000   1.0  ...    1.000000   \n",
       "\n",
       "         Tumor_cells  T_cell_CD4  Plasma_cell       cDC1       cDC2  \\\n",
       "CDH11       1.000000    1.000000     1.000000   1.000000   1.000000   \n",
       "DOCK9       1.000000    1.000000     1.000000   1.000000   1.000000   \n",
       "CLIC3       1.000000    1.000000     1.000000   1.000000   1.000000   \n",
       "CCNG2       1.000000    1.000000     1.000000  52.746868  56.873905   \n",
       "PCDH17      1.000000    1.000000     1.000000   1.000000   1.000000   \n",
       "...              ...         ...          ...        ...        ...   \n",
       "CDKN3      93.431375    1.000000     1.000000   1.000000   1.000000   \n",
       "CAMK4       1.000000  213.361118     1.000000   1.000000   1.000000   \n",
       "WIF1        1.000000    1.000000     1.000000   1.000000   1.000000   \n",
       "TMEM156     1.000000   76.778009   119.800078   1.000000   1.000000   \n",
       "SLCO5A1     1.000000    1.000000     1.000000   1.000000   1.000000   \n",
       "\n",
       "             B_cell  Neutrophils   DC_mature  T_cell_regulatory  \n",
       "CDH11      1.000000          1.0    1.000000           1.000000  \n",
       "DOCK9      1.000000          1.0    1.000000           1.000000  \n",
       "CLIC3      1.000000          1.0    1.000000           1.000000  \n",
       "CCNG2      1.000000          1.0  314.848934           1.000000  \n",
       "PCDH17     1.000000          1.0    1.000000           1.000000  \n",
       "...             ...          ...         ...                ...  \n",
       "CDKN3      1.000000          1.0    1.000000           1.000000  \n",
       "CAMK4      1.000000          1.0    1.000000         106.219895  \n",
       "WIF1       1.000000          1.0    1.000000           1.000000  \n",
       "TMEM156  167.259363          1.0    1.000000           1.000000  \n",
       "SLCO5A1    1.000000          1.0  272.951785           1.000000  \n",
       "\n",
       "[1149 rows x 24 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "o_sig_matrix = pd.read_csv(\"/home/biolab/Projects/LAMs_wd/data/signature_matrix_sq_24_cell_type_renamed.csv\", index_col=0)\n",
    "o_sig_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aa0f5d67-28a6-4a64-b481-d32c2265261b",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f\"/home/biolab/Projects/LAMs_wd/data/luad_macrophage_spots_cell2location_spatial_model_SC_Y_sig_matrix_adeno_LUSC_patients_{env}.txt\", \"r\") as i:\n",
    "    spots = i.read().split(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9a99932e-fbba-40b0-96e8-e293aca677c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "497"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(spots)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "db17cd38-1059-4388-b55c-d6faaabb75fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['spaceranger110_count_36210_OTAR_LNGsp9476042_GRCh38-2020-A_ACTGTAGCACTTTGGA-1',\n",
       " 'spaceranger110_count_36210_OTAR_LNGsp9476043_GRCh38-2020-A_TCGTGTCACGCTGACA-1',\n",
       " 'spaceranger110_count_36210_OTAR_LNGsp9476042_GRCh38-2020-A_TCGGAGAGTATCGGGA-1',\n",
       " 'spaceranger110_count_36210_OTAR_LNGsp9476042_GRCh38-2020-A_TTCTAACCGAAGCTTA-1',\n",
       " 'spaceranger110_count_36210_OTAR_LNGsp9476045_GRCh38-2020-A_CCATTTCTACCTATTA-1',\n",
       " 'spaceranger110_count_36210_OTAR_LNGsp9476045_GRCh38-2020-A_CCCAGTTAAGGCGCCG-1',\n",
       " 'spaceranger110_count_36210_OTAR_LNGsp9476045_GRCh38-2020-A_TAAACCCAGGAGGGCA-1',\n",
       " 'spaceranger110_count_36210_OTAR_LNGsp9476043_GRCh38-2020-A_ACCGACACATCTCCCA-1',\n",
       " 'spaceranger110_count_36210_OTAR_LNGsp9476045_GRCh38-2020-A_GCGGTTCCCTATCATG-1',\n",
       " 'spaceranger110_count_38262_OTAR_LNGsp10206167_GRCh38-2020-A_GCGTCTCTGCATTGGG-1']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spots[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9dad690-9347-4b3e-9079-d3136214b013",
   "metadata": {},
   "source": [
    "Run Cell2location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "12fae158-00e5-42d3-b290-2ccf7fb5e5a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_cell2loc(ifolder, environment):\n",
    "    seed = 42\n",
    "    torch.manual_seed(seed)\n",
    "    device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "    outDir = f\"/home/biolab/Projects/LAMs_wd/results/cell2location/spatial_model_[SC_Y_sig_matrix_adeno_macrophages_sub_spots]_[LUSC_patients]/\"\n",
    "    regDir = f\"/home/biolab/Projects/LAMs_wd/data/E-MTAB-13530/{ifolder}/cell2location/regression_model/\"\n",
    " \n",
    "\n",
    "    inf_aver = None\n",
    "    # Export estimated expression signature in each cluster\n",
    "    \n",
    "    inf_aver = o_sig_matrix.copy()\n",
    "\n",
    "    # Scale up by average sample scaling factor. This corrects for sequencing depth\n",
    "    # inf_aver = inf_aver * adata_sc_model.uns[\"mod\"][\"post_sample_means\"][\"sample_scaling\"].mean() # Need to check if this is still necessary\n",
    "\n",
    "    # Pick up Visium\n",
    "    if env == \"Tumour\":\n",
    "        sample_names = lung_config[environment][\"spatial\"][\"LUSC\"]\n",
    "    else:\n",
    "        sample_names = lung_config[environment][\"spatial\"]\n",
    "    print(sample_names)\n",
    "    \n",
    "    ####\n",
    "    #assign you path here, where your patients' data are located\n",
    "    ############################\n",
    "    inDir = \"/home/biolab/Projects/LAMs_wd/data/E-TAMB-13530/\"\n",
    "    ########################################################################\n",
    "    \n",
    "    slides = [read_and_qc(sample_name, path=inDir) for sample_name in sample_names]\n",
    "    slides_ = []\n",
    "    \n",
    "    #now here i select a subset of spots\n",
    "    for slide in slides:\n",
    "        slides_.append(slide[list(set(slide.obs_names).intersection(spots))])\n",
    "    slides = slides_\n",
    "    \n",
    "    print(slides)\n",
    "    \n",
    "    \n",
    "    adata_vis = slides[0].concatenate(slides[1:], batch_key=\"sample\", uns_merge=\"unique\", batch_categories=sample_names, index_unique=None)\n",
    "\n",
    "    # Mitochondria-encoded (MT) genes should be removed for spatial mapping\n",
    "    # Find mitochondria-encoded (MT) genes\n",
    "    adata_vis.var[\"MT_gene\"] = [gene.startswith(\"MT-\") for gene in adata_vis.var[\"SYMBOL\"]]\n",
    "    # Remove MT genes for spatial mapping (keeping their counts in the object)\n",
    "    adata_vis.obsm[\"MT\"] = adata_vis[:, adata_vis.var[\"MT_gene\"].values].X.toarray()\n",
    "    adata_vis = adata_vis[:, ~adata_vis.var[\"MT_gene\"].values]\n",
    "    # Reset the index for compatibility with the scRNA-seq counts\n",
    "    adata_vis.var.set_index(\"SYMBOL\", inplace=True)\n",
    "\n",
    "    # Find shared genes and subset both anndata and reference signatures\n",
    "    adata_vis.var_names_make_unique()\n",
    "    \n",
    "    inf_aver_index = np.array(inf_aver.index, dtype=str)\n",
    "    adata_vis_var_names = np.array(adata_vis.var_names, dtype=str)\n",
    "    intersect = np.intersect1d(adata_vis_var_names, inf_aver_index)\n",
    "    #intersect = np.intersect1d(adata_vis.var_names, inf_aver.index)\n",
    "    #adata_vis = adata_vis[:, intersect] # This doesn't work -- think there's some incompatibility between ScanPy and AnnData versions. Use the line below instead\n",
    "    adata_vis = adata_vis[:, adata_vis.var.index.isin(intersect)].copy()\n",
    "    inf_aver = inf_aver.loc[intersect, :]\n",
    "    # Reindex for cell2location\n",
    "    inf_aver = inf_aver.reindex(adata_vis.var.index)\n",
    "\n",
    "\n",
    "    # Training cell2location\n",
    "    cell2location.models.Cell2location.setup_anndata(adata=adata_vis, batch_key=\"sample\")\n",
    "    mod = cell2location.models.Cell2location(adata_vis,\n",
    "                                             cell_state_df=inf_aver,\n",
    "                                             detection_alpha=200, # Controls the normalisation of the within-experiment variation of the RNA detection (default value for now)\n",
    "                                             N_cells_per_location=8\n",
    "                                             )\n",
    "    '''\n",
    "                                             **model_kwargs={\n",
    "                                                # Prior on the number of cells, cell types and co-located groups. Hyperparameter inputs go here\n",
    "                                                \"cell_number_prior\": {\n",
    "                                                    # - N - the expected number of cells per location:\n",
    "                                                    \"cells_per_spot\": 8,\n",
    "                                                    # - A - the expected number of cell types per location:\n",
    "                                                    \"factors_per_spot\": 9,\n",
    "                                                    # - Y - the expected number of co-located cell type groups per location\n",
    "                                                    \"combs_per_spot\": 5\n",
    "                                                },\n",
    "                                                # Prior beliefs on the sensitivity of spatial technology:\n",
    "                                                \"gene_level_prior\": {\n",
    "                                                    # Prior on the mean\n",
    "                                                    \"mean\": 0.5,\n",
    "                                                    # Prior on standard deviation\n",
    "                                                    # A good choice of this value should be at least 2 times lower that the mean\n",
    "                                                    \"sd\": 0.15\n",
    "                                                }\n",
    "                                             }\n",
    "                                             )\n",
    "    '''\n",
    "\n",
    "    # Try 2500 batches initially, rather than training on the entire data-set\n",
    "    mod.train(batch_size=2500, train_size=1, max_epochs=3000)\n",
    "    mod.save(f\"{outDir}{environment}\", overwrite=True)\n",
    "    adata_vis = mod.export_posterior(adata_vis, sample_kwargs={'num_samples': 1000, 'batch_size': 2500})\n",
    "    adata_vis.write(f\"{outDir}{environment}/sp_our_sig_matrix_macrophage_cts_{environment}.h5ad\")\n",
    "\n",
    "    train_performance = mod.history[\"elbo_train\"]\n",
    "    train_performance.plot(logy=True)\n",
    "    plt.xlabel(\"Training epochs\")\n",
    "    plt.ylabel(\"-ELBO loss\")\n",
    "    plt.title(f\"Lung ({environment})\")\n",
    "    plt.legend(loc=\"best\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f\"{outDir}{environment}/ELBO_Cell2location_{environment}.png\")\n",
    "    plt.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b727bb29-107e-4657-b439-d99d72a5c6c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['spaceranger110_count_36210_OTAR_LNGsp9476045_GRCh38-2020-A', 'spaceranger110_count_38262_OTAR_LNGsp10206168_GRCh38-2020-A', 'spaceranger110_count_38262_OTAR_LNGsp10206167_GRCh38-2020-A', 'spaceranger110_count_36210_OTAR_LNGsp9476042_GRCh38-2020-A', 'spaceranger110_count_36210_OTAR_LNGsp9476043_GRCh38-2020-A']\n",
      "[View of AnnData object with n_obs × n_vars = 116 × 36601\n",
      "    obs: 'in_tissue', 'array_row', 'array_col', 'sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'mt_frac'\n",
      "    var: 'feature_types', 'genome', 'SYMBOL', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'mt'\n",
      "    uns: 'spatial'\n",
      "    obsm: 'spatial', View of AnnData object with n_obs × n_vars = 78 × 36601\n",
      "    obs: 'in_tissue', 'array_row', 'array_col', 'sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'mt_frac'\n",
      "    var: 'feature_types', 'genome', 'SYMBOL', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'mt'\n",
      "    uns: 'spatial'\n",
      "    obsm: 'spatial', View of AnnData object with n_obs × n_vars = 94 × 36601\n",
      "    obs: 'in_tissue', 'array_row', 'array_col', 'sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'mt_frac'\n",
      "    var: 'feature_types', 'genome', 'SYMBOL', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'mt'\n",
      "    uns: 'spatial'\n",
      "    obsm: 'spatial', View of AnnData object with n_obs × n_vars = 104 × 36601\n",
      "    obs: 'in_tissue', 'array_row', 'array_col', 'sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'mt_frac'\n",
      "    var: 'feature_types', 'genome', 'SYMBOL', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'mt'\n",
      "    uns: 'spatial'\n",
      "    obsm: 'spatial', View of AnnData object with n_obs × n_vars = 104 × 36601\n",
      "    obs: 'in_tissue', 'array_row', 'array_col', 'sample', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'mt_frac'\n",
      "    var: 'feature_types', 'genome', 'SYMBOL', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'mt'\n",
      "    uns: 'spatial'\n",
      "    obsm: 'spatial']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True (cuda), used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "HPU available: False, using: 0 HPUs\n",
      "You are using a CUDA device ('NVIDIA RTX A4000') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 3000/3000: 100%|██████████| 3000/3000 [03:56<00:00, 12.57it/s, v_num=1, elbo_train=2.59e+5]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`Trainer.fit` stopped: `max_epochs=3000` reached.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 3000/3000: 100%|██████████| 3000/3000 [03:56<00:00, 12.68it/s, v_num=1, elbo_train=2.59e+5]\n",
      "Sampling local variables, batch: 100%|██████████| 1/1 [00:13<00:00, 13.42s/it]\n",
      "Sampling global variables, sample: 100%|██████████| 999/999 [00:13<00:00, 71.74it/s]\n"
     ]
    }
   ],
   "source": [
    "run_cell2loc(\"dataset\", env)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "339203b5-70d5-47ef-8327-a9be7c96a563",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (jupyter_env)",
   "language": "python",
   "name": "jupyter_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
